File: Problems/Algorithms/98. Validate Binary Search Tree/valid_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def aux(root: Optional[TreeNode], min_value: int, max_value: int) -> bool:
            if root is None:
                return True

            if root.val > min_value and root.val < max_value:
                return aux(root.left, min_value, root.val) and aux(root.right, root.val, max_value)
            return False

        return aux(root, -2e31, 2e31)

File: Problems/Algorithms/98. Validate Binary Search Tree/valid_bst.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn is_valid_bst(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        fn aux(root: Option<Rc<RefCell<TreeNode>>>, min_value: i64, max_value: i64) -> bool {
            match root {
                None => true,
                Some(node) => {
                    let node = node.borrow();
                    if (node.val as i64) > min_value && (node.val as i64) < max_value {
                        return aux(node.left.clone(), min_value, node.val as i64) && aux(node.right.clone(), node.val as i64, max_value);
                    }
                    false
                },
            }
        }

        aux(root, -2147483649, 2147483648)
    }
}

